### PR TITLE
Feature/manual shipment tracking

### PR DESCRIPTION
**Pull request by pradeepdeveloper143**
Please do check everything and let me know.

### CODE REVIEW

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+		add_filter( 'wc_order_statuses', [$this,'hez_add_shipped_to_order_statuses'] );
+
+		// Add Custom fields for woocommerce order
+		add_action( "add_meta_boxes", [$this,"hez_add_meta_boxes"] );
+		
+		// Ajax handler function
+		add_action( 'wp_ajax_hezarfen_shipping_data_action', [$this,'hezarfen_shipping_data_action'] );
+
+		// Add Custom fields for woocommerce order
+		add_action( 'wp_ajax_nopriv_hezarfen_shipping_data_action', [$this,'hezarfen_shipping_data_action'] );
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:85]: we don't need it since we don't have support for not logged-in users, please remove this line.

https://developer.wordpress.org/reference/hooks/wp_ajax_action/

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	            $new_order_statuses['wc-hez_shipped'] = 'Kargolandı';
+	        }
+	    }
+	    return $new_order_statuses;
+	}
+	// Add Custom fields for woocommerce order
+	public function hez_add_meta_boxes()
+	{
+		if(isset($_GET['post'])){
+			$order_id=$_GET['post'];
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:123]: Please sanitize the data

https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+		    add_meta_box(
+		        "woocommerce-order-my-custom",
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:125]: the metabox id should be specific for the task, please give an meaningful meta box id

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+		        __( "Shipping Options" ),
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:126]: Please add **hezarfen-for-woocommerce** text-domain as second argument

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+		        "side",
+		        "default"
+		    );		
+		}		
+	}
+	// Add Custom fields for woocommerce order
+	public function hez_order_shipping_options(){
+	    global $woocommerce,$wpdb;
+	    $shipping_carriers=$this->shipping_carriers;
+	    $order_id=$_GET['post'];
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:138]: there is need to sanitization here

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+			    </select>
+			</div>
+			
+		    <div class="wp-clearfix">
+		    	<label>Shipping Tracking Code</label><br>
+		    	<input type="text" name="hez_ship_track_no"  value="<?php echo $hez_ship_track_no; ?>" class="shipping_fields hez_ship_track_no">	
+		    </div>
+		    <br>
+		    <div class="text-right">
+		    	<button type="button" name="update_shipping" class="button button-primary update-shipping">Update Shipping</button>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:181]: Please make translatable the **Update Shipping** string

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	        });
+	    </script>
+		<?php
+	}
+
+	// Store Custom fields for woocommerce order
+	public function hezarfen_shipping_data_action(){
+		global $wpdb;
+		$order_id=$_POST['id'];
+		update_post_meta( $order_id, 'hez_ship_carrier', $_POST['hez_ship_carrier'] );
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:230]: sanitize the data

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+		update_post_meta( $order_id, 'hez_ship_track_no', $_POST['hez_ship_track_no'] );
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:231]: again, data sanitization needed on here

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+	    ) );
+	}
+
+	// Add custom status to order status list
+	public function hez_add_shipped_to_order_statuses( $order_statuses ) {
+	    $new_order_statuses = array();
+	    foreach ( $order_statuses as $key => $status ) {
+	        $new_order_statuses[ $key ] = $status;
+	        if ( 'wc-processing' === $key ) {
+	            $new_order_statuses['wc-hez_shipped'] = 'Kargolandı';
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:114]: The 'Kargolandı' string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+	            });
+	        });
+	    </script>
+		<?php
+	}
+
+	// Store Custom fields for woocommerce order
+	public function hezarfen_shipping_data_action(){
+		global $wpdb;
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:228]: There should be nonce verification on here

https://codex.wordpress.org/WordPress_Nonces#Verifying_a_nonce_passed_in_an_AJAX_request

**Review by mskapusuz — CHANGES_REQUESTED** [review]
in addition these, after the "update shipping" button click, the page is being refreshed. It's unexpected since there is ajax support on there

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+			    display: inline-block;
+			    line-height: 2;
+			    padding: 0 10px;
+			    border-radius: 3px;
+			}
+		</style>
+		<h3>Tracking Details</h3>
+		<table class="table">
+			<tr>
+				<th>Shipping Carrier</th>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:275]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+				<th>Tracking link</th>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:276]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+			    border-color: #2271b1;
+			    color: #fff;
+			    text-decoration: none;
+			    display: inline-block;
+			    line-height: 2;
+			    padding: 0 10px;
+			    border-radius: 3px;
+			}
+		</style>
+		<h3>Tracking Details</h3>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:272]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+			    {
+			    	if($shipping_carrier['is_available'])
+			    		echo "<option value='".$key."' ".($hez_ship_carrier==$key?'selected':'').">".$shipping_carrier['name']."</option>";
+			    }
+			    ?>
+			    </select>
+			</div>
+			
+		    <div class="wp-clearfix">
+		    	<label>Shipping Tracking Code</label><br>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:176]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+			}
+			.error_input{
+				border: 1px solid #d63638 !important;
+			}
+		</style>
+		<form action="" method="post" >
+		    <div class="wp-clearfix">
+			    <label>Shipping Carrier</label><br>
+			    <select name="hez_ship_carrier" class="shipping_fields hez_ship_carrier">
+			    	<option value="">Select Shippiing Carrier</option>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:164]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+			.text-right{
+				text-align: right;
+			}
+			.error_input{
+				border: 1px solid #d63638 !important;
+			}
+		</style>
+		<form action="" method="post" >
+		    <div class="wp-clearfix">
+			    <label>Shipping Carrier</label><br>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:162]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+
+	// Function for register custom order status
+	public function hez_register_shipped_order_status() {
+	    register_post_status( 'wc-hez_shipped', array(
+	        'label'                     => 'Kargolandı',
+	        'public'                    => true,
+	        'exclude_from_search'       => false,
+	        'show_in_admin_all_list'    => true,
+	        'show_in_admin_status_list' => true,
+	        'label_count'               => _n_noop( 'Kargolandı <span class="count">(%s)</span>', 'Kargolandı <span class="count">(%s)</span>' )
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:104]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	private function get_shipping_providers(){
+		return apply_filters( 'hezarfen_supported_shipping_carriers', $this->shipping_carriers );
+	}
+
+	// Function for register custom order status
+	public function hez_register_shipped_order_status() {
+	    register_post_status( 'wc-hez_shipped', array(
+	        'label'                     => 'Kargolandı',
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:99]: the string should be translatable.

**Review by mskapusuz — CHANGES_REQUESTED** [review]

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 1 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:85]: we don't need it since we don't need ajax handler support for not logged-in users, please remove this line.

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 18 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:99]: the label string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+	        'public'                    => true,
+	        'exclude_from_search'       => false,
+	        'show_in_admin_all_list'    => true,
+	        'show_in_admin_status_list' => true,
+	        'label_count'               => _n_noop( 'Kargolandı <span class="count">(%s)</span>', 'Kargolandı <span class="count">(%s)</span>' )
+	    ) );
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:105]: the "Kargolandı" strings should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+	public function hez_add_shipped_to_order_statuses( $order_statuses ) {
+	    $new_order_statuses = array();
+	    foreach ( $order_statuses as $key => $status ) {
+	        $new_order_statuses[ $key ] = $status;
+	        if ( 'wc-processing' === $key ) {
+	            $new_order_statuses['wc-hez_shipped'] = 'Kargolandı';
+	        }
+	    }
+	    return $new_order_statuses;
+	}
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:118]: I think, there is no need a if condition and foreach on here.

Do you have any special reason to don't make like in the following:

`public function hez_add_shipped_to_order_statuses( $order_statuses ) {
     $order_statuses['wc-hez_shipped'] = 'Kargolandı';
    return $order_statuses;
}`

I couldn't understand the relation between wc-processing and our custom status(wc-hez_shipped)

please let me know if I'm missing something.

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 2 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:123]: Please sanitize the data

https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 5 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:138]: Please sanitize the data

https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+		<form action="" method="post" >
+		    <div class="wp-clearfix">
+			    <label>Shipping Carrier</label><br>
+			    <select name="hez_ship_carrier" class="shipping_fields hez_ship_carrier">
+			    	<option value="">Select Shippiing Carrier</option>
+			    <?php 
+			    foreach($shipping_carriers as $key=>$shipping_carrier)
+			    {
+			    	if($shipping_carrier['is_available'])
+			    		echo "<option value='".$key."' ".($hez_ship_carrier==$key?'selected':'').">".$shipping_carrier['name']."</option>";
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:169]: option value and option label should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 14 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:176]: the label should be translatable

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 6 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:181]: the button label should be translatable

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+		                     type : "POST",
+		                     dataType : "json",
+		                     url : "<?php echo home_url();?>/wp-admin/admin-ajax.php",
+		                     data : {action: "hezarfen_shipping_data_action",id:order_id,hez_ship_carrier:hez_ship_carrier,hez_ship_track_no:hez_ship_track_no},
+		                     success: function(response) {
+		                           $('.update-shipping').removeAttr('disabled');
+		                           $('.hez_ship_track_no').removeClass('error');
+		                           $('.hez_ship_carrier').removeClass('error');
+		                           location.reload();
+
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:213]: we don't need location.reload since we've AJAX support. We should avoid hard reload

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 13 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:272]: the string should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php (hunk 12 above)
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:276]: the strings should be translatable.

File: includes/packages/Manuel-Shipment/manual-shipment.php [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+				<th>Tracking link</th>
+			</tr>
+			<tr>
+				<td >
+					<?php 
+					if($shipping_carriers[$hez_ship_carrier]['logo']!=''){?>
+						<img src="<?php echo $shipping_carriers[$hez_ship_carrier]['logo']; ?>" class="img-logo">
+					<?php 
+					}
+					 echo '<span class="shipping_carrier_title">'.$shipping_carriers[$hez_ship_carrier]['name'].'</span>'; ?>
```
Comment (mskapusuz) [inline includes/packages/Manuel-Shipment/manual-shipment.php:285]: use escape before print the variable

**Review by mskapusuz — CHANGES_REQUESTED** [review]
